zproject/backends.py

```diff
@@ -1500,6 +1509,91 @@ def get_verified_emails(self, *args: Any, **kwargs: Any) -> List[str]:
+                             'sv_SE', 'th_TH', 'tr_TR', 'uk_UA', 'vi_VI',
+                             'zh_CN']
+        for locale in supported_locales:
+            if django_language_code in locale:
+                return locale
+        return 'en_US'
+
+    # Below we override appropriate methods of the superclass to adapt them to our needs.
+    def get_or_create_state(self) -> str:
+        '''
+        Creates the Oauth2 state parameter in first step of the flow, before redirecting
+        the user to the IdP. We're dealing with Apple here, which will send the user back
+        to us with a POST request. Normally, we rely on being able to store certain parameters
+        in the user's session and use them after the redirect. Session is a Lax cookie, so
+        it won't be sent by the user's browser in the POST redirect from Apple. To get around this,
+        we will store the parameters in redis under a key derived from the state. That will allow us
+        to validate the state and retrieve the params after the redirect - by querying redis for
+        the key derived from the state sent in the POST redirect.
+        '''
+        request_data = self.strategy.request_data().dict()
+        data_to_store = {
+            key: request_data[key] for key in self.standard_relay_params
+            if key in request_data
+        }
+
+        # Generate a random string of 32 alphanumeric characters.
+        state = self.state_token()
+        put_dict_in_redis(redis_client, 'apple_auth_{token}',
+                          data_to_store, self.REDIS_EXPIRATION_SECONDS,
+                          token=state)
+        return state
+
+    # We override this to retrieve the data stored in redis and store it
+    # in session along with validating the state.
+    def validate_state(self) -> Optional[str]:
+        """
+        Oauth2 state validation, done when the user is redirected to us back from the Apple.
+        This overrides the upstream method, because we need
+        to retrieve certain parameters for this auth attempt that we stored in the first step
+        of the flow under a key derived from the state.
+        """
+        request_state = self.get_request_state()
+
+        if not request_state:
+            self.logger.info("Sign in with Apple failed: missing state parameter.")
+            raise AuthMissingParameter(self, 'state')
+
+        formatted_request_state = "apple_auth_" + request_state
+        redis_data = get_dict_from_redis(redis_client, "apple_auth_{token}",
+                                         formatted_request_state)
+        if redis_data is None:
+            self.logger.info("Sign in with Apple failed: bad state token.")
+            raise AuthStateForbidden(self)
+
+        for param, value in redis_data.items():
+            if param in self.standard_relay_params:
+                self.strategy.session_set(param, value)
+        return request_state
+
```

Comment:
We need to include `decode_id_token` override, with the native-flow logic stripped away from it. It's important, because we fix some flaws from upstream there - not catching some exception, and raising an exception incorrectly. We should have those fixes in this PR.